src/main/java/org/ossgang/commons/monads/AsyncMaybe.java

```diff
@@ -202,8 +184,8 @@ protected AsyncMaybe(CompletableFuture<T> future, Function<ThrowingSupplier<T>,
      * @return A successful AsyncMaybe the new value if the transformation succeeds. An unsuccessful AsyncMaybe otherwise.
      */
     public <R> AsyncMaybe<R> map(ThrowingFunction<T, R> function) {
-        requireNonNull(function);
-        return AsyncMaybe.fromCompletableFuture(future.thenApplyAsync(v -> requireNonNull(uncheckedFunction(function).apply(v), NULL_VALUE_MSG), ASYNC_MAYBE_POOL));
+        requireNonNull(function, "Mapping function cannot be null");
+        return new AsyncMaybe<>(stage.thenApplyAsync(stageResult -> stageResult.map(requireNonNullResult(function)::apply), ASYNC_MAYBE_POOL));
```

Comment:
I think Maybe::map should already ensure that the value is not null - is this just to give a better error message?

Reply:
Yes, indeed it does. Here was to have the same error everywhere :slightly_smiling_face: